lib/ansible/modules/packaging/os/mas.py

```diff
@@ -0,0 +1,290 @@
+
+    def check_signin(self):
+        ''' Verifies that the user is signed in to the Mac App Store '''
+
+        # Only check this once per execution
+        if self._checked_signin:
+            return
+
+        rc, out, err = self.run(['account'])
+        if out.split("\n", 1)[0].rstrip() == 'Not signed in':
+            self.module.fail_json(msg='You must be signed in to the Mac App Store')
+
+        self._checked_signin = True
+
+    def exit(self):
+        ''' Exit with the data we have collected over time '''
+
+        msgs = []
+        if self.count_install > 0:
+            msgs.append('Installed {0} app(s)'.format(self.count_install))
+        if self.count_upgrade > 0:
+            msgs.append('Upgraded {0} app(s)'.format(self.count_upgrade))
+        if self.count_uninstall > 0:
+            msgs.append('Uninstalled {0} app(s)'.format(self.count_uninstall))
+
+        if msgs:
+            self.result['changed'] = True
+            self.result['msg'] = ', '.join(msgs)
+
+        self.module.exit_json(**self.result)
+
+    def get_current_state(self, command):
+        ''' Returns the list of all app IDs; command can either be 'list' or 'outdated' '''
+
+        rc, raw_list, err = self.run([command])
+        rows = raw_list.split("\n")
+        list = {}
+        for r in rows:
+            # Format: "123456789 App Name"
+            r = r.split(' ', 1)
+            if len(r) == 2:
+                list[r[1]] = int(r[0])
+
+        return list
+
+    def installed(self):
+        ''' Returns the list of installed apps '''
+
+        # Populate cache if not already done
+        if self._installed is None:
+            self._installed = self.get_current_state('list')
+
+        return self._installed.values()
+
+    def is_installed(self, id):
+        ''' Checks whether the given app is installed '''
+
+        return int(id) in self.installed()
+
+    def is_outdated(self, id):
+        ''' Checks whether the given app is installed, but outdated '''
+
+        return int(id) in self.outdated()
+
+    def outdated(self):
+        ''' Returns the list of installed, but outdated apps '''
+
+        # Populate cache if not already done
+        if self._outdated is None:
+            self._outdated = self.get_current_state('outdated')
+
+        return self._outdated.values()
+
+    def run(self, cmd):
+        ''' Runs a command of the `mas` tool '''
+
+        cmd.insert(0, self.mas_path)
+        return self.module.run_command(cmd, False)
+
+    def upgrade_all(self):
+        ''' Upgrades all installed apps and sets the correct result data '''
+
+        outdated = self.outdated()
+
+        if not self.module.check_mode:
+            self.check_signin()
+
+            rc, out, err = self.run(['upgrade'])
+            if rc != 0:
+                self.module.fail_json(msg='Could not upgrade all apps: ' + out.rstrip())
+
+        self.count_upgrade += outdated
```

Comment:
Here, `self.count_upgrade` is an int, and `outdated` (as above) a `dict` resp. `list`.

Reply:
You are right! Fixed it in https://github.com/mheap/ansible/pull/3.